packages/tools/bundleDefaultPlugins.ts

```diff
@@ -0,0 +1,93 @@
+import { execCommand2 } from './tool-utils';
+import { pathExists, mkdir, readFile, move, remove, writeFile } from 'fs-extra';
+import { DefaultPluginsInfo } from '@joplin/lib/services/plugins/PluginService';
+import getDefaultPluginsInfo from '@joplin/lib/services/plugins/defaultPlugins/desktopDefaultPluginsInfo';
+const fetch = require('node-fetch');
+
+interface PluginAndVersion {
+   [pluginId: string]: string;
+}
+
+interface PluginIdAndName {
+	[pluginId: string]: string;
+}
+
+export const localPluginsVersion = async (defaultPluginDir: string, defaultPluginsInfo: DefaultPluginsInfo): Promise<PluginAndVersion> => {
+	if (!await pathExists(join(defaultPluginDir))) await mkdir(defaultPluginDir);
+	const localPluginsVersions: PluginAndVersion = {};
+
+	for (const pluginId of Object.keys(defaultPluginsInfo)) {
+
+		if (!await pathExists(join(defaultPluginDir, pluginId))) {
+			localPluginsVersions[pluginId] = '0.0.0';
+			continue;
+		}
+		const data = await readFile(`${defaultPluginDir}/${pluginId}/manifest.json`, 'utf8');
+		const manifest = JSON.parse(data);
+		localPluginsVersions[pluginId] = manifest.version;
+	}
+	return localPluginsVersions;
+};
+
+async function downloadFile(url: string, outputPath: string) {
+	const response = await fetch(url);
+	await writeFile(outputPath, await response.buffer());
+}
+
+export async function extractPlugins(currentDir: string, defaultPluginDir: string, downloadedPluginsNames: PluginIdAndName) {
+	for (const pluginId of Object.keys(downloadedPluginsNames)) {
+		await execCommand2(`tar xzf ${currentDir}/${downloadedPluginsNames[pluginId]}`);
+		await move(`package/publish/${pluginId}.jpl`,`${defaultPluginDir}/${pluginId}/plugin.jpl`, { overwrite: true });
+		await move(`package/publish/${pluginId}.json`,`${defaultPluginDir}/${pluginId}/manifest.json`, { overwrite: true });
+		await remove(`${downloadedPluginsNames[pluginId]}`);
+		await remove('package');
+	}
+}
+
+export const downloadPlugins = async (localPluginsVersions: PluginAndVersion, defaultPluginsInfo: DefaultPluginsInfo, manifests: any): Promise<any> => {
+
+	const downloadedPluginsNames: PluginIdAndName = {};
+	for (const pluginId of Object.keys(defaultPluginsInfo)) {
+		if (localPluginsVersions[pluginId] === defaultPluginsInfo[pluginId].version) continue;
+		const response = await fetch(`https://registry.npmjs.org/${manifests[pluginId]._npm_package_name}`);
+
+		if (!response.ok) {
+			const responseText = await response.text();
+			throw new Error(`Cannot download plugin file ${responseText.substr(0,500)}`);
+		}
+		const releaseText = await response.text();
+		const release = JSON.parse(releaseText);
+
+		const pluginUrl = release.versions[defaultPluginsInfo[pluginId].version].dist.tarball;
+
+		const pluginName = `${manifests[pluginId]._npm_package_name}-${defaultPluginsInfo[pluginId].version}.tgz`;
+		await downloadFile(pluginUrl, pluginName);
+
+		if (!await pathExists(pluginName)) throw new Error(`${pluginName} cannot be downloaded`);
```

Comment:
You don't need this check, because downloadFile should throw if there's an error

Reply:
Yes. Thanks for catching this.